3.2/root/usr/share/container-scripts/mongodb/init-petset-replset.sh

```diff
@@ -0,0 +1,146 @@
+'
+readonly WAIT_SECONDARY_STATUS="
+  var mbrs;
+  while (!mbrs || mbrs.length == 0 || mbrs[0].state != 2) {
+    printjson(rs.status());
+    sleep(1000);
+    mbrs = rs.status().members.filter(function(el) {
+      return el.name.indexOf(\"${MEMBER_HOST}:\") > -1;
+    });
+  };
+  print(mbrs[0].stateStr);
+"
+
+# Outputs available endpoints (hostnames) to stdout.
+# This also includes hostname of the current pod.
+#
+# Uses the following global variables:
+# - MONGODB_SERVICE_NAME (optional, defaults to 'mongodb')
+function find_endpoints() {
+    local service_name="${MONGODB_SERVICE_NAME:-mongodb}"
+
+    # Extract host names from lines like this: "10 33 0 mongodb-2.mongodb.myproject.svc.cluster.local."
+    dig "${service_name}" SRV +search +short | cut -d' ' -f4 | rev | cut -c2- | rev
+}
+
+# Performs initialization of the primary node.
+#
+# Arguments:
+# - $1 hostname
+#
+# Uses the following global variables:
+# - MONGODB_REPLICA_NAME
+# - MONGODB_ADMIN_PASSWORD
+# - WAIT_PRIMARY_STATUS
+function setup_primary() {
+    local member_host="${1}"
+
+    local endpoints
+    endpoints="$(find_endpoints)"
+
+    if [ "$(echo "${endpoints}" | grep -v "^${member_host}" | wc -l)" != '0' ]; then
+        info "Skipping replica set initialization (already exist)"
+        return
+    fi
+
+    local config="{_id: \"${MONGODB_REPLICA_NAME}\", members: [ {_id: 0, host: \"${member_host}\"} ]}"
+
+    info "Initiating MongoDB replica using: ${config}"
+    mongo admin --eval "rs.initiate(${config});${WAIT_PRIMARY_STATUS}" --quiet
+
+    info "Creating MongoDB users ..."
+    mongo_create_admin
+    mongo_create_user "-u admin -p ${MONGODB_ADMIN_PASSWORD}"
+
+    info "Successfully initialized replica set"
+}
+
+# Adds a host to the replica set configuration. It is safe to call this function
+# if the host is already in the configuration.
+#
+# Arguments:
+# - $1: host address[:port]
+#
+# Global variables:
+# - MONGODB_REPLICA_NAME
+# - MONGODB_ADMIN_PASSWORD
+# - SLEEP_TIME
+# - WAIT_SECONDARY_STATUS
+function add_member() {
+  local host="$1"
+  info "Adding ${host} to replica set ..."
+
+  local script
+  script="
+    var ret = rs.add('${host}');
+    if (ret.ok) {
+      quit(0);
+    }
+    // ignore error if host is already in the configuration
+    if (ret.code == 163) {
+      quit(0);
+    }
+    printjson(ret);
+    quit(1);
+  "
+
+  # TODO: replace this with a call to `replset_addr` from common.sh, once it returns host names.
+  local replset_addr
+  replset_addr="${MONGODB_REPLICA_NAME}/$(find_endpoints | paste -s -d,)"
+
+  if ! mongo admin -u admin -p "${MONGODB_ADMIN_PASSWORD}" --host "${replset_addr}" --eval "${script}" --quiet; then
+      info "ERROR: couldn't join replica set!"
+      return 1
+  fi
+
+  info "Successfully joined replica set"
+  info "Waiting for becoming a SECONDARY node ..."
+
+  # REVIEW: why do we need to wait for "SECONDARY" status? What if this member
+  # joins and becomes the PRIMARY?
```

Comment:
@php-coder why do we need to wait for "SECONDARY" status? What if this member joins and becomes the PRIMARY?